metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala

```diff
@@ -642,4 +511,310 @@ object BuildTargets {
+    def isSourceRoot: java.util.Set[AbsolutePath]
+    // if workspace contains symlinks, original source items are kept here and source items dealiased
+    def originalSourceItems: java.util.Set[AbsolutePath]
+
+    def targetToConnection
+        : scala.collection.Map[BuildTargetIdentifier, BuildServerConnection]
+    def sourceBuildTargets(
+        sourceItem: AbsolutePath
+    ): Option[Iterable[BuildTargetIdentifier]]
+
+    def allTargetRoots: Iterator[AbsolutePath]
+    def all: Iterator[BuildTarget]
+    def allBuildTargetIds: Seq[BuildTargetIdentifier]
+    def allScala: Iterator[ScalaTarget]
+    def allJava: Iterator[JavaTarget]
+    def scalaTarget(id: BuildTargetIdentifier): Option[ScalaTarget]
+    def javaTarget(id: BuildTargetIdentifier): Option[JavaTarget]
+
+    def targetRoots(buildTarget: BuildTargetIdentifier): List[AbsolutePath]
+    def javaTargetRoot(buildTarget: BuildTargetIdentifier): Option[AbsolutePath]
+    def scalaTargetRoot(
+        buildTarget: BuildTargetIdentifier
+    ): Option[AbsolutePath]
+
+    def info(id: BuildTargetIdentifier): Option[BuildTarget]
+    def targetJarClasspath(
+        id: BuildTargetIdentifier
+    ): Option[List[AbsolutePath]]
+    def targetClasspath(id: BuildTargetIdentifier): Option[List[String]]
+    def targetClassDirectories(id: BuildTargetIdentifier): List[String]
+
+    def allWorkspaceJars: Iterator[AbsolutePath]
+
+    def addSourceItem(
+        sourceItem: AbsolutePath,
+        buildTarget: BuildTargetIdentifier
+    ): Unit
+  }
+
+  final case class DataSeq(head: WritableData, tail: List[WritableData]) {
+    private lazy val list: List[Data] = head :: tail
+    def iterator: Iterator[Data] = list.iterator
+    def writableDataIterator: Iterator[WritableData] = (head :: tail).iterator
+    def iterable: Iterable[Data] = list.toIterable
+
+    def fromIterators[T](f: Data => Iterator[T]): Iterator[T] =
+      iterator.flatMap(f)
+    def fromOptions[T](f: Data => Option[T]): Option[T] =
+      fromIterators(f(_).iterator).take(1).toList.headOption
+  }
+
+  trait WritableData extends Data {
+    def linkSourceFile(id: BuildTargetIdentifier, source: AbsolutePath): Unit
+    def reset(): Unit
+    def addWorkspaceBuildTargets(result: WorkspaceBuildTargetsResult): Unit
+    def addScalacOptions(result: ScalacOptionsResult): Unit
+    def addJavacOptions(result: JavacOptionsResult): Unit
+    def addDependencySource(
+        sourcesJar: AbsolutePath,
+        target: BuildTargetIdentifier
+    ): Unit
+    def resetConnections(
+        idToConn: List[(BuildTargetIdentifier, BuildServerConnection)]
+    ): Unit
+    def onCreate(source: AbsolutePath): Unit
+  }
+
+  object Data {
+    def create(): WritableData =
+      new DataImpl
+  }
+
+  private final class DataImpl extends WritableData {
+
+    val sourceItemsToBuildTarget
+        : TrieMap[AbsolutePath, ConcurrentLinkedQueue[BuildTargetIdentifier]] =
```

Comment:
Side grumble, I wish we could change our scalafmt config to not allow breaking on `:` like this.